### PR TITLE
Fix android open link verify

### LINKED ISSUE #1397: [Feature Request] Enable ADB-Like Direct URL Opening in App Functionality for ﻿openLink
With ADB, it's possible to open URLs directly with the app, even when the URLs are not verified.

```
adb shell am start -a android.intent.action.VIEW -d "https://www.company.com/invitation/c77ca9d0-5ff7-4b59-8d1b-4802b054e9cd" com.company.android.maestro.debug
```

This request is to extend the functionality of `openLink` to replicate this behavior.

**Describe the solution you'd like**

Add a property (e.g., `openWithViewIntent`) to `openLink` that allows opening the URL in the same way as ADB does:

```yaml
-  openLink:
    link: https://www.company.com/invitation/c77ca9d0-5ff7-4b59-8d1b-4802b054e9cd
    autoVerify: true
    openWithViewIntent: true
```

**Describe alternatives you've considered**

An alternative solution currently is to add steps in your flow to add the URLs to the "Open by default" list. However, this approach increases the runtime of the flow and may be more fragile.

```yaml
# Adding company.com and www.company.com as supported links on Android
-  runFlow:
    when:
      platform: Android
    commands:
      - launchApp: "com.android.settings"
      - tapOn: "Apps"
      - tapOn: "All apps"
      - scrollUntilVisible:
          element:
            text: "App Name"
      - tapOn: "App Name"
      - scrollUntilVisible:
          element:
            text: "Open by default"
      - tapOn: "Open by default"
      - tapOn: "Add link"
      - tapOn: "company.com"
      - tapOn: "www.company.com"
      - tapOn: "Add"

-  openLink:
    link: https://www.company.com/invitation/c77ca9d0-5ff7-4b59-8d1b-4802b054e9cd
    autoVerify: true
```

**Additional context**

Discussion on Slack: [Slack Discussion](https://mobile-dev-inc.slack.com/archives/C041FU72T54/p1693299687056699)

### LINKED ISSUE #1956: Android exception while opening a deeplink - java.lang.NullPointerException: appName must not be null
### Is there an existing issue for this?

- [X] I have searched the existing issues and didn't find mine.

### Steps to reproduce

1) Installed the android react native app on emulator
2) tried to open an app deeplink with following steps as suggested by Maestro official doc

Code Block:
```
appId: com.example.app
---
- launchApp:
    clearState: true
- openLink:
    link: 'https://link-stage.test.fan/test123'
- assertVisible: 'Welcome'
```

Please Note: It seems to run fine on iOS simulator

### Actual results

failed with the message:
java.lang.NullPointerException: appName must not be null
[maestro test maestro/deeplink.yaml] 	at maestro.drivers.AndroidDriver.autoVerifyWithAppName(AndroidDriver.kt:533)
[maestro test maestro/deeplink.yaml] 	at maestro.drivers.AndroidDriver.autoVerifyApp(AndroidDriver.kt:519)
[maestro test maestro/deeplink.yaml] 	at maestro.drivers.AndroidDriver.openLink(AndroidDriver.kt:513)
[maestro test maestro/deeplink.yaml] 	at maestro.Maestro.openLink(Maestro.kt:487)
[maestro test maestro/deeplink.yaml] 	at maestro.orchestra.Orchestra.openLinkCommand(Orchestra.kt:701)
[maestro test maestro/deeplink.yaml] 	at maestro.orchestra.Orchestra.executeCommand(Orchestra.kt:262)
[maestro test maestro/deeplink.yaml] 	at maestro.orchestra.Orchestra.executeCommands(Orchestra.kt:201)
[maestro test maestro/deeplink.yaml] 	at maestro.orchestra.Orchestra.runFlow(Orchestra.kt:111)
[maestro test maestro/deeplink.yaml] 	at maestro.orchestra.Orchestra.runFlow$default(Orchestra.kt:75)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.runner.MaestroCommandRunner.runCommands(MaestroCommandRunner.kt:185)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.runner.TestRunner$runSingle$result$1.invoke(TestRunner.kt:54)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.runner.TestRunner$runSingle$result$1.invoke(TestRunner.kt:51)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.runner.TestRunner.runCatching(TestRunner.kt:145)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.runner.TestRunner.runSingle(TestRunner.kt:51)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.command.TestCommand$handleSessions$1$1$results$1$1$1.invoke(TestCommand.kt:291)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.command.TestCommand$handleSessions$1$1$results$1$1$1.invoke(TestCommand.kt:249)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.session.MaestroSessionManager.newSession(MaestroSessionManager.kt:107)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.session.MaestroSessionManager.newSession$default(MaestroSessionManager.kt:58)
[maestro test maestro/deeplink.yaml] 	at maestro.cli.command.TestCommand$handleSessions$1$1$results$1$1.invokeSuspend(TestCommand.kt:249)
[maestro test maestro/deeplink.yaml] 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:585)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:802)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:706)
[maestro test maestro/deeplink.yaml] 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:693)
[maestro test maestro/deeplink.yaml] ==== Debug output (logs & screenshots) ====

### Expected results

The app should have opened the deeplink successfully

### About app

cross-platform app, react native

### About environment

Device: Pixel 3a
OS: API 33
MacOS: Sonoma - 14.6.1

### Logs

<details>
<summary>Logs</summary>

```
<!-- Replace this line with your logs. *DO NOT* remove the backticks! -->
```

</details>

### Maestro version

1.37.9

### How did you install Maestro?

install script (https://get.maestro.mobile.dev)

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by tokou**
### TL;DR 

I fixed a whole lot of stuff that was wrong with the `autoVerify`. It should now work as expected.

> [!NOTE]
> I cherry picked the commit from #1998 locally as I seemed to stumble on that same issue in other places as I tested

### Long version

- I started investigating this #1956 issue. It was pretty straightforward to reproduce and fix.
  - The origin is due to the fact that the reading of the app name to tap on the right cell of the Chooser is done through `apkMeta.name` 
  - This value which contains the value defined in `<application android:label="name">` in the Manifest. Which obviously is not mandatory, not guaranteed to be there, and even better, not what the chooser shows as value
  - The fix was simply to check the nullability of the value and throw if it's nullable (because it's a `String!` from Java-world)
  - However, while doing this, I noticed the code related to `autoVerify` and I opened another can of worms...
- `autoVerify` is supposed to do 2 things. Select the right app from the Chooser and validate some of the onboarding screens from Chrome. Both had issues in the way they were implemented
  - Chrome onboarding
    - Depending on the version of Android and Chrome that you have, there are different variations of up to 4 onboarding screens
    - I just fixed the function to handle all of those
  - Selection of the wanted app
    - Before API 31 (Android 12) you always get a Chooser if your app has declared an IntentFilter with ACTION_VIEW on the domain
      - I fixed the code that selects the right app from the Chooser
      - I kept the selection of "Just once" over "Always" even though I've hesitated to switch it 🤔 
      - This is still not perfect as we ultimately would need to get the result of something like 
`intent.resolveActivity(packageManager)` in order to get the labels associated with the right Activities
    - Starting from API 31, there's no more Chooser but there's the whole "Open by default" settings menu
      - I added handling for all of that automatically when `autoVerify` is selected
      - I did not reverse the setting after the link is opened. I think it makes more sense like that I think. Maybe we should align "Just once" in the Chooser? 

## Testing

This is interesting to test. I'm not sure how we can automate that (or if we should). It would make a lot of integration tests. How were the various drivers tested?

### What I did

- First I had a very simple test flow
```
appId: dev.mobile.maestro.app
---
- openLink:
    link: 'https://maestro.mobile.dev'
    autoVerify: true
    browser: false
- openLink:
    link: 'https://maestro.mobile.dev'
    autoVerify: true
    browser: true
```
- I created a new module with a very barebones app `dev.mobile.maestro.app` that I used to test the null name
```
    <application>
        <activity
            android:name=".MainActivity"
            android:exported="true"
            >
            <intent-filter>
                <category android:name="android.intent.category.LAUNCHER"/>
                <action android:name="android.intent.action.MAIN"/>
            </intent-filter>
            <intent-filter>
                <action android:name="android.intent.action.VIEW" />
                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />
                <data android:scheme="https" />
                <data android:host="maestro.mobile.dev" />
            </intent-filter>
        </activity>
    </application>
```
```
class MainActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(TextView(this).apply { text = "${applicationInfo.name}" })
    }
}
```
- I created a whole bunch of brand new emulators (API 24, 25, 26, 28, 32, 33, 34, 35)
- And then I ran the flow over and over on all the emulators until everything ran as expected
- In order to get the different Chrome screens, you have to Force stop and Clear data for it and they are a bit random

### Chrome

| Accept | Sign in | Sync | Notifications |
|-------|------|------|------|
| chrome_welcome_accept | chrome_welcome_account | chrome_sync | chrome_notifications |

### Chooser

| New | App | Browser |
|-----|-----|-----|
| Screenshot_20240902_034915 | Screenshot_1725245194 | Screenshot_1725245113 |

### Open by default

| Before | After |
|------|------|
| Screenshot_1725244809 | Screenshot_1725244813 |

## Issues fixed

Fixes #1956 (Definitely fixed)
Fixes #1397 (Fixed in spirit but not in letter. I think it's just better to use `autoVerify` for that purpose and not add another option)

### CODE REVIEW

**Comment by bartekpacia** [comment]
Whoah, this is awesome. Thanks!

> I created a new module with a very barebones app `dev.mobile.maestro.app` that I used to test the null name

Could you share this code? We could add this to [our e2e testing suite](https://github.com/mobile-dev-inc/maestro/tree/main/e2e):
- add app binary link to [manifest](https://github.com/mobile-dev-inc/maestro/blob/3ec7cac71a3b115ccbf583bfe696e0abf4e4070a/e2e/manifest.txt)
- add tests to [workspaces](https://github.com/mobile-dev-inc/maestro/tree/3ec7cac71a3b115ccbf583bfe696e0abf4e4070a/e2e/workspaces)/<your_app_name>

Alternatively, the relevant intent filter stuff could be added to our [demo_app](https://github.com/mobile-dev-inc/demo_app). Wdyt?


> I created a whole bunch of brand new emulators (API 24, 25, 26, 28, 32, 33, 34, 35)

This is a action item for us to run on all supported API levels in CI. Currently [we run only on API 28](https://github.com/mobile-dev-inc/maestro/blob/3ee902e2c8d5d9c9bdf84906d562cc23376cb5c8/.github/workflows/test-e2e.yaml#L44).

File: maestro-client/src/main/java/maestro/android/AndroidAppFiles.kt [hunk 1]
```diff
@@ -38,11 +38,15 @@ object AndroidAppFiles {
...
+        val apkPath = dadb.shell("pm path $appId").output.removePrefix("package:").trim()
         val dst = File.createTempFile("tmp", ".apk")
-        dadb.pull(dst, apkPath)
+        try {
+            dadb.pull(dst, apkPath)
+        } catch (e: IOException) {
+            val newApkPath = "/sdcard/$appId.apk"
+            dadb.shell("cp $apkPath $newApkPath")
+            dadb.pull(dst, newApkPath)
+        }
```
Comment (bartekpacia) [inline maestro-client/src/main/java/maestro/android/AndroidAppFiles.kt:49]: Why this change? I'm curious to learn why `/sdcard` is used as fallback

**Review by bartekpacia — APPROVED** [review]

**Comment by bartekpacia** [comment]
This topic is much more complex that I though. Thank you very much for sharing your experience and contributing.

If you think that [the documentation for `openLink`](https://maestro.mobile.dev/api-reference/commands/openlink) could be expanded, we would greatly appreciate that as well.

File: maestro-client/src/main/java/maestro/android/AndroidAppFiles.kt (hunk 1 above)
Comment (tokou) [reply maestro-client/src/main/java/maestro/android/AndroidAppFiles.kt:49]: - I encountered a `permission denied` when pulling the APK on some of the API levels
- This is because `/data` is not accessible to pull from without the device being rooted in those cases
- However running `shell cp` and copying the files over is possible
- And `/sdcard` is accessible to pull from

See this [SO answer](https://stackoverflow.com/questions/41150038/cannot-pull-apk-from-android-device-with-adb/43157127#43157127) for example

If we add multi-API-level testing on the CI, we should add simple tests for `AndroidAppFiles` as this is the kind of things that breaks with API changes.

**Comment by ubuntudroid** [comment]
Why was this reverted? 🤔 Looks like a super useful bunch of changes and I was looking forward to finally get rid of the custom steps we were doing in our tests. Were there breaking issues?

**Comment by amanjeetsingh150** [comment]
Hey folks, just an FYI on this PR. This created a regression in openlink command with auto verify flag on cloud. While I think this can be reworked and fixed. I'll do a root cause on what exactly caused issue and we can take in rest of the changes and fix rest. I'll do a follow up here this week
